File: backend/src/main/java/org/luxons/sevenwonders/game/moves/CardFromHandMove.java
package org.luxons.sevenwonders.game.moves;

import java.util.List;

import org.luxons.sevenwonders.game.api.PlayerMove;
import org.luxons.sevenwonders.game.api.Table;
import org.luxons.sevenwonders.game.cards.Card;

public abstract class CardFromHandMove extends Move {

    CardFromHandMove(int playerIndex, Card card, PlayerMove move) {
        super(playerIndex, card, move);
    }

    @Override
    public boolean isValid(Table table, List<Card> playerHand) {
        return playerHand.contains(getCard());
    }

}

File: sw-engine/src/main/kotlin/org/luxons/sevenwonders/game/moves/CardFromHandMove.kt
package org.luxons.sevenwonders.game.moves

import org.luxons.sevenwonders.game.PlayerContext
import org.luxons.sevenwonders.game.api.PlayerMove
import org.luxons.sevenwonders.game.cards.Card

internal abstract class CardFromHandMove(move: PlayerMove, card: Card, player: PlayerContext) :
    Move(move, card, player) {

    init {
        if (!player.hand.contains(card)) {
            throw InvalidMoveException(this, "card '${card.name}' not in hand")
        }
    }
}
